### PR TITLE
Replaced old language service with Roslyn-based one

### PR DESCRIPTION
**Pull request by otawfik-ms**
@Microsoft/fsharp-compiler 
This will stay in the new branch (roslyn), and users can try it by installing the vsix produced from the CI build there.
The current version just loads the new language service, and has no UI features (bare text editor).

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/extension.vsixmanifest [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
-  </Identifier>
-  <References />
-  <Content>
-    <MefComponent>FSharp.Editor.dll</MefComponent>
-  </Content>
-</Vsix>
\ No newline at end of file
+  </Installation>
+  <Dependencies>
+    <Dependency Id="Microsoft.Framework.NDP" DisplayName="Microsoft .NET Framework" d:Source="Manual" Version="[4.5,)" />
```
Comment (Jetski5822) [inline vsintegration/src/FSharp.Editor/extension.vsixmanifest]: Does 4.5 here mean anything?

File: vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    [<Literal>]
+    let packageGuid = "871D2A70-12A2-4e42-9440-425DD92A4116"
+    [<Literal>]
+    let languageServiceGuid = "BC6DD5A5-D4D6-4dab-A00D-A51242DBAF1B"
+    [<Literal>]
+    let editorFactoryGuid = "4EB7CCB7-4336-4FFD-B12B-396E9FD079A9"
+    [<Literal>]
+    let codePageEditorFactoryGuid = "82A16493-EF43-47E0-B42D-D87BAAB5335D"
+    [<Literal>]
+    let svsSettingsPersistenceManagerGuid = "9B164E40-C3A2-4363-9BC5-EB4039DEF653"
```
Comment (cston) [inline vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs]: Perhaps add `String` suffix to these constants.

File: vsintegration/src/FSharp.Editor/FSharpProjectSiteService.fs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+open System.Composition
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Host
+open Microsoft.CodeAnalysis.Host.Mef
+
+open Microsoft.VisualStudio.FSharp.LanguageService
+open Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+
+// Currently the CompilationOptions type in Roslyn in sealed and there's no way to set the compilation options for a project.
```
Comment (cston) [inline vsintegration/src/FSharp.Editor/FSharpProjectSiteService.fs]: `is sealed`

File: vsintegration/src/FSharp.Editor/extension.vsixmanifest [hunk 4]
```diff
@@ -1,24 +1,23 @@
...
-    <Locale>1033</Locale>
-    <SupportedProducts>
-      <VisualStudio Version="14.0">
-        <Edition>Pro</Edition>
-        <Edition>VWDExpress</Edition>
-        <Edition>WDExpress</Edition>
-      </VisualStudio>
-    </SupportedProducts>
-    <SupportedFrameworkRuntimeEdition MinVersion="4.5" />
+<!-- Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
```
Comment (cston) [inline vsintegration/src/FSharp.Editor/extension.vsixmanifest]: Is this the correct copyright notice?

Comment (cston) [inline vsintegration/src/FSharp.Editor/extension.vsixmanifest]: `Microsoft Corporation`?

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs [hunk 5]
```diff
@@ -3,618 +3,119 @@
...
 
+open Microsoft.FSharp.Compiler.SourceCodeServices
 
-                
-/// This class defines capabilities of the language service. 
-/// CodeSense = true\false, for example
-type internal FSharpLanguagePreferences(site, langSvc, name) = 
-    inherit LanguagePreferences(site, langSvc, name) 
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.SolutionCrawler
```
Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: Are there any public types in the `SolutionCrawler` namespace?

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs [hunk 6]
```diff
@@ -3,618 +3,119 @@
...
-            // Invalidate the configuration if we notice any add for any DependencyFiles 
-            let checkOptions = ProjectSitesAndFiles.GetProjectOptionsForProjectSite(projectSite, "")
-            this.FSharpChecker.InvalidateConfiguration(checkOptions)
 
-        member this.DependencyFileChanged (filename) = 
-            this.NotifyFileTypeCheckStateIsDirty filename
+                let projectId = workspace.ProjectTracker.GetOrCreateProjectIdForPath(site.ProjectFileName(), site.ProjectFileName())
+                if obj.ReferenceEquals(workspace.ProjectTracker.GetProject(projectId), null) then
+                    let projectSite = new FSharpProjectSite(hier, this.SystemServiceProvider, workspace, site.ProjectFileName());
+                    projectSite.Initialize(hier, site)
```
Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: Consider extracting a local for `site.ProjectFileName()`.

File: vsintegration/src/FSharp.Editor/extension.vsixmanifest (hunk 1 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.Editor/extension.vsixmanifest]: Changing to 4.6 to align to Roslyn's binaries

File: vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs (hunk 2 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs]: :thumbsup:

File: vsintegration/src/FSharp.Editor/FSharpProjectSiteService.fs (hunk 3 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.Editor/FSharpProjectSiteService.fs]: :thumbsup:

File: vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs (hunk 2 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/LanguageServiceUtils.fs]: Code style in other projects is to leave it to type inference. Any reason to change it?

File: vsintegration/src/FSharp.Editor/extension.vsixmanifest (hunk 4 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.Editor/extension.vsixmanifest]: :+1:

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs (hunk 5 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: And this is why we need sort/remove unused opens :)

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs (hunk 6 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: :+1:

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            Seq.iter(fun file -> this.AddDocument(hier, file))
+        
+        this.GetReferences(site.CompilerFlags()) |> 
+            Seq.iter(fun ref -> this.AddReference(ref) |> ignore)
+        
+        // Capture the F# specific options that we'll pass to the type checker.
+        checkOptions <- Some(ProjectSitesAndFiles.GetProjectOptionsForProjectSite(site, site.ProjectFileName()))
+
+    member this.GetReferences(flags : string[]) =
+        let (|Reference|_|) (f : string) = if f.StartsWith("-r:") then Some (f.Replace("-r:", "")) else None
```
Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: `f.Substring(3)`

Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: :+1:

Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: Or perhaps:
```let references = flags |> Seq.where(fun flag -> flag.StartsWith("-r:")) |> Seq.map(fun flag -> flag.Substring(3))```

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs [hunk 8]
```diff
@@ -3,618 +3,119 @@
...
-                    let defines = this.ProjectSitesAndFiles.GetDefinesForFile(rdt, filename)
-                    let sourceTokenizer = FSharpSourceTokenizer(defines,filename)
-                    sourceTokenizer.CreateLineTokenizer(source))
-
-            let colorizer = new FSharpColorizer(this.CloseColorizer, buffer, scanner) 
-            (buffer :?> IVsUserData).SetData(&guid, colorizer) |> ErrorHandler.ThrowOnFailure |> ignore
-            colorizer
-        | _ -> colorizerObj :?> FSharpColorizer
+    override this.ContentTypeName = FSharpCommonConstants.FSharpContentTypeName
+    override this.GetFormattedTextChanges(_, _, _, _) = System.Collections.Generic.List<Text.TextChange>() :> System.Collections.Generic.IList<Text.TextChange>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: `Array.empty :> _` or `upcast Array.empty`

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs [hunk 9]
```diff
@@ -3,618 +3,119 @@
...
-                            | :? IVsTextLines as tl ->
-                                ls.ProjectSitesAndFiles.UnsetSource(tl)
-                            | _ -> ()
-            finally 
-                if IntPtr.Zero <> unkdoc then Marshal.Release(unkdoc)|>ignore
-            VSConstants.S_OK
+        [|
+            editorFactory :> IVsEditorFactory;
+            codePageEditorFactory :> IVsEditorFactory;
+        |] :> IEnumerable<IVsEditorFactory>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: `seq<_>` or `seq<IVSEditorFactory>`

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        this.ProjectTracker.AddProject(this)
+
+        site.AdviseProjectSiteChanges(FSharpCommonConstants.FSharpLanguageServiceCallbackName,
+                                        new AdviseProjectSiteChanges(fun () -> this.OnProjectSettingsChanged(hier, site)))
+
+        site.AdviseProjectSiteClosed(FSharpCommonConstants.FSharpLanguageServiceCallbackName, 
+                                        new AdviseProjectSiteChanges(fun () -> this.Disconnect()))
+
+        // Add files and references.
+        site.SourceFilesOnDisk() |>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: AFAIR `Seq.iter` is not marked as `inline` meaning that this call will allocate `FSharpFunc` object. Any reasons to prefer it to plain `for file in site.SourceFilesOnDisk() do this.Add.Document(hier, file)`?

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        this.GetReferences(site.CompilerFlags()) |> 
+            Seq.iter(fun ref -> this.AddReference(ref) |> ignore)
+        
+        // Capture the F# specific options that we'll pass to the type checker.
+        checkOptions <- Some(ProjectSitesAndFiles.GetProjectOptionsForProjectSite(site, site.ProjectFileName()))
+
+    member this.GetReferences(flags : string[]) =
+        let (|Reference|_|) (f : string) = if f.StartsWith("-r:") then Some (f.Replace("-r:", "")) else None
+
+        let references = flags |>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: you can adjust active pattern + use a `Seq.choose`:
```
let (|Reference|_|) (f : string) = if f.StartsWith("-r:") && f.Length > 3 then Some (f.Substring 3) else None
let references = flags |> Seq.choose (|Reference|_|)

```

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            | (VSConstants.S_OK, id) -> id
+            | _ -> uint32 VSConstants.VSITEMID.Nil
+
+        let document = this.ProjectTracker.DocumentProvider.TryGetDocumentForFile(this, itemid, file, SourceCodeKind.Regular, fun x -> true)
+        this.AddDocument(document, true)
+
+    member internal this.OnProjectSettingsChanged(hier: IVsHierarchy, site : IProjectSite) = 
+        let sourceFiles = site.SourceFilesOnDisk()
+        // Added files.
+        sourceFiles |>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: 1. commonly accepted format for pipe operator is
```
a
|> b
|> c
```

2. IMO there are really no reasons to prefer Seq.iter to `for..in`

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        // Added files.
+        sourceFiles |> 
+            Seq.where(fun file -> not(this.ContainsFile(file))) |> 
+            Seq.iter(fun file -> this.AddDocument(hier, file))
+        // Removed files.
+        this.GetCurrentDocuments() |> 
+            Seq.where(fun doc -> not(sourceFiles |> Seq.contains(doc.FilePath))) |>
+            Seq.iter(fun doc -> this.RemoveDocument(doc))
+
+        let references = this.GetReferences(site.CompilerFlags())
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: currently `this.GetReferences` returns result of `seq.where` which is deferred sequence. This sequence will be traversed several times making reasoning about performance quite complicated (see below)

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        // Added references
+        references |>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: first traversal

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            Seq.iter(fun doc -> this.RemoveDocument(doc))
+
+        let references = this.GetReferences(site.CompilerFlags())
+        // Added references
+        references |> 
+            Seq.where(fun ref -> not(this.HasMetadataReference(ref))) |>
+            Seq.iter(fun ref -> this.AddReference(ref) |> ignore)
+        // Removed references
+        this.GetCurrentMetadataReferences() |>
+            Seq.where(fun ref -> not(references |> Seq.contains(ref.FilePath))) |>
```
Comment (vladima) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: traversal on every call to `Seq.contains`

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    member internal this.OnProjectSettingsChanged(hier: IVsHierarchy, site : IProjectSite) = 
+        let sourceFiles = site.SourceFilesOnDisk()
+        // Added files.
+        sourceFiles |> 
+            Seq.where(fun file -> not(this.ContainsFile(file))) |> 
+            Seq.iter(fun file -> this.AddDocument(hier, file))
+        // Removed files.
+        this.GetCurrentDocuments() |> 
+            Seq.where(fun doc -> not(sourceFiles |> Seq.contains(doc.FilePath))) |>
+            Seq.iter(fun doc -> this.RemoveDocument(doc))
```
Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: Calling `RemoveDocument` while iterating over `GetCurrentDocuments` is fragile since it assumes that `GetCurrentDocuments` is returning a copy. Instead, consider looping over `GetCurrentDocuments` to create a list of files to remove, then looping over that list calling `RemoveDocument`.

Comment (cston) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: Alternatively, the signature of `GetCurrentDocuments` could be changed to return ```ImmutableArray<T>```.

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 11 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: :+1:

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs (hunk 8 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: :+1:

File: vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs (hunk 9 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpLanguageService.fs]: :+1:

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        artifacts.Value
+        
+    member this.Preferences = 
+        if this.Unhooked then raise Error.UseOfUnhookedLanguageServiceState
+        if not this.IsInitialized then raise Error.UseOfUninitializedLanguageServiceState
+        preferences.Value
+        
+    member this.SourceFactory = 
+        if this.Unhooked then raise Error.UseOfUnhookedLanguageServiceState
+        if not this.IsInitialized then raise Error.UseOfUninitializedLanguageServiceState
```
Comment (cston) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: Perhaps extract `CheckHooked` and `CheckInitialized` methods.

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+            sourceFactory <- None
+            serviceProvider <- None
+    
+    /// Respond to project settings changes
+    member this.OnProjectSettingsChanged(site:IProjectSite) = 
+        // The project may have changed its references.  These would be represented as 'dependency files' of each source file.  Each source file will eventually start listening
+        // for changes to those dependencies, at which point we'll get OnDependencyFileCreateOrDelete notifications.  Until then, though, we just 'make a note' that this project is out of date.
+        bgRequests.AddOutOfDateProjectFileName(site.ProjectFileName()) 
+        for filename in site.SourceFilesOnDisk() do
+            let rdt = this.ServiceProvider.RunningDocumentTable
```
Comment (cston) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: Move `let rdt` outside of `for`.

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 10 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: No preference here, it is following the same pattern found in other files. Will change it for the noted optimization reason.

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+    /// Do OnIdle processing for the whole language service. dirtyForTypeCheckFiles can be set by events 
+    /// raised on the background compilation thread.  
+    member this.OnIdle() = 
+        for file in dirtyForTypeCheckFiles do
+            let rdt = this.ServiceProvider.RunningDocumentTable
+            match this.ProjectSitesAndFiles.TryGetSourceOfFile(rdt, file) with
+            | Some source -> source.RecordChangeToView()
+            | None ->  ()
+        dirtyForTypeCheckFiles <- Set.empty
```
Comment (cston) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: If `dirtyForTypeCheckFiles` can be set on the background thread, is there a race condition here?

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+        let mutable colorizerObj = null
+        
+        (buffer :?> IVsUserData).GetData(&guid, &colorizerObj) |> ignore
+        match colorizerObj with
+        | null ->
+            let scanner = 
+                new FSharpScanner(fun source ->
+                    // Note: in theory, the next few lines do not need to be recomputed every line.  Instead we could just cache the tokenizer
+                    // and only update it when e.g. the project system notifies us there is an important change (e.g. a file rename, etc).
+                    // In practice we have been there, and always screwed up some non-unit-tested/testable corner-cases.
```
Comment (cston) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: Perhaps "and always missed some ...".

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 12 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: as other comment

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs (hunk 17 above)
Comment (otawfik-ms) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: This class is basically exposing functionality of the old language service. I'll be deleting that code iteratively as I expose the new features. The whole class is copied as it is from the old language service in order not to break existing tests.

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs (hunk 18 above)
Comment (otawfik-ms) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: please check above comment

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs (hunk 19 above)
Comment (otawfik-ms) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: please check above comment

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs (hunk 20 above)
Comment (otawfik-ms) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: please check above comment

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 13 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: Not using a deferred sequence anymore.

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 14 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: please check above comment

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 15 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: please check above comment

File: vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs (hunk 16 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs]: :+1:

**Comment by cston** [comment]
LGTM

**Comment by otawfik-ms** [comment]
thanks @cston and @vladima !

File: vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs (hunk 17 above)
Comment (dsyme) [inline vsintegration/tests/Salsa/FSharpLanguageServiceTestable.fs]: Good approach!

**Comment by dsyme** [comment]
This is such a great start - super to see this.
